### PR TITLE
[asset owners] API and graphQL support

### PR DESCRIPTION
**Pull request by clairelin135**
This PR supports:
- defining asset owners as a list of strings on `@asset`, `AssetOut`, and `AssetSpec`
- fetching an asset's owners via graphQL
- marking asset owners as experimental

Not in scope:
- Querying for all assets with a given owner. We're not sure what the search looks like yet, and would likely want a unified approach to query across multiple attributes (owner, column schema, etc.)
- Defining asset owners on graph-backed assets. This is lower priority and not in scope for MVP.

### CODE REVIEW

**Comment by clairelin135** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #19616** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19616?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19616?utm_source=stack-comment).

File: python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py [hunk 1]
```diff
@@ -2054,3 +2054,24 @@ def _specified_elsewhere() -> int:
             return my_op()
 
         assert _specified_elsewhere  # appease linter
+
+
+def test_asset_owners():
```
Comment (rexledesma) [inline python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py:2059]: Can we add a test case against `owners=[]`?

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 2]
```diff
@@ -315,6 +318,12 @@ def __init__(
         )
 
         self._is_subset = check.bool_param(is_subset, "is_subset")
+        self._owners_by_key = check.opt_mapping_param(
+            owners_by_key, "owners_by_key", key_type=AssetKey, value_type=list
+        )
+
+        if self._owners_by_key and any(self._owners_by_key.values()):
```
Comment (rexledesma) [inline python_modules/dagster/dagster/_core/definitions/assets.py:325]: nit: consolidate check
```suggestion
        if any(self._owners_by_key.values()):
```

**Review by rexledesma — COMMENTED** [review]
Looks fine, just need one more test

**Review by sryza — COMMENTED** [review]
I think we should make this experimental until we have the chance to get some feedback on it. Otherwise, lgtm.

**Comment by clairelin135** [comment]
@sryza this PR currently throws an experimental warning if owners are defined

**Review by sryza — COMMENTED** [review]
Ah, oop,s missed that. We now have an `@experimental_param` decorator – can we use that? Also, the docstring should mention that the parameter is experimental (although I noticed that we don't do that with other params on `AssetSpec` and `AssetOut`, so maybe best to leave it out there until we can add it to all of them).

**Review by rexledesma — APPROVED** [review]

**Comment by clairelin135** [comment]
@sryza @rexledesma could I get another look at this?

This PR now limits owner strings to be emails or strings with the `team:` prefix. Hoping to get this into this week's release.

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 3]
```diff
@@ -69,6 +72,19 @@
 ASSET_SUBSET_INPUT_PREFIX = "__subset_input__"
 
 
+@whitelist_for_serdes
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/assets.py:75]: I'm hesitant to commit to storage types that we need to support forever, before we've learned more about what the ontology should look like. There's also the question of `UserAssetOwner` vs. `UserOwner` vs. `User` vs. `Person`. Thoughts on just storing a string for now?

Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/assets.py:75]: Yeah that makes sense, we can remove the `whitelist_for_serdes` tags here.

I'd prefer on the backend to still have `UserAssetOwner` and `TeamAssetOwner` though so we don't need to check for each string whether its a user, team, or if the `team:` prefix is inside. But these objects we can change whenever we want.

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/assets.py:75]: 👍

**Review by sryza — APPROVED** [review]

**Pull request merged**